---
title: "Allelic Imbalance Analysis"
output: html_notebook
---
```{r}
#!/gsc/software/linux-x86_64-centos7/R-4.1.3/lib64/R/bin/R
# R 4.1.3
# x86_64-centos7-linux-gnu
library(plyranges)
library(VariantAnnotation)
library(MutationalPatterns)
library(genomation)
library(Gviz)
library(BSgenome.Hsapiens.UCSC.hg38)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(systemPipeR)
library(biomaRt)
library(LDlinkR)
library(AllelicImbalance)
library(broom)
library(gt)
library(gtsummary)
library(UpSetR)
library(tidyverse)

const_canonical_chromosomes <- str_c("chr", c(1:22, "X"))

vector_marks <- c("H3K4me3", "H3K27ac", "H3K4me1", "H3K9me3", "H3K27me3", "H3K36me3")
vector_celltypes <- c("BC", "LP", "LC", "SC")
vector_individuals <- c("Individual_14_17_LP_DNA",
                        "Individual_14_18_BC__356K_RNA",
                        "Individual_11_18_BC_RNA_DNA",
                        "Individual_22_18_BC_RNA_DNA",
                        "Individual_24_18_BC_RNA_DNA",
                        "Individual_38_18_LC_RNA_DNA",
                        "Individual_30_18_LC_RNA_DNA",
                        "Individual_15_18_LC_RNA_DNA")
```

Get Allelic Imbalance on CEMT_154 H3K4me3 BC on alt-allele-driven (positive) cis-hBTLs - TEST SINGLE RUN
```{r}
# Initialize results tibble
ase_depths <- tibble()

# Import ChIP-seq treatment vcf file (test)
ase_readcounts_H3K4me3_BC_Individual_1 <- 
  readVcfAsVRanges("ase/output/test/test2.vcf.gz") %>%
  as.data.frame() %>%
  select(-width, -strand) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE, ignore.strand = TRUE, seqinfo = seqinfo(BSgenome.Hsapiens.UCSC.hg38)) %>%
  keepSeqlevels(const_canonical_chromosomes, pruning.mode = "coarse")

# Scan VCF for heterozygotic variant calls and keep only cis-hBTLs at heterozygotic positions for that individual
# These are the positions that need to be tested for the particular individual-modality
ase_heterozygotePositions_H3K4me3_CEMT_154 <-
  join_overlap_intersect(consensus.all.spaas.rpkm.bed %>% 
                           plyranges::filter(mark == "H3K4me3", celltype == "BC", spaa == TRUE) %>% 
                           keepSeqlevels(const_canonical_chromosomes, pruning.mode = "coarse"),
                         readVcfAsVRanges("/projects/epigenomics3/epigenomics3_results/users/ahauduc/arp/project/modules/1/output2/H3K4me3/CEMT_154/wgs/CEMT_154.recal_hard_filtered_phased_snvs.vcf.gz") %>%
                           as.data.frame() %>%
                           select(-width, -strand) %>%
                           makeGRangesFromDataFrame(keep.extra.columns = TRUE, ignore.strand = TRUE, seqinfo = seqinfo(BSgenome.Hsapiens.UCSC.hg38)) %>%
                           keepSeqlevels(const_canonical_chromosomes, pruning.mode = "coarse") %>% 
                           plyranges::filter(AF == 0.5)) %>% 
  plyranges::select(-everything())

# Subset the ChIP-seq treatment VCF with proper heterozygotic, modality-relevant positions
# And select only the positions with some evidence of heterozygotic alignment
ase_readcounts_H3K4me3_BC_Individual_1_subsetted <-
  ase_readcounts_H3K4me3_BC_Individual_1 %>% 
  join_overlap_intersect(ase_heterozygotePositions_H3K4me3_CEMT_154) %>% 
  plyranges::filter(refDepth != 0 & altDepth != 0)

# Pull out the vectors of depths where ref or alt != 0
current_tibble <-
  tibble(mark = "H3K4me3",
         celltype = "BC",
         individual = "1",
         ref_depth_mean = ase_readcounts_H3K4me3_BC_Individual_1_subsetted$refDepth %>% mean(),
         alt_depth_mean = ase_readcounts_H3K4me3_BC_Individual_1_subsetted$altDepth %>% mean())

# Add to tibble
ase_depths <- bind_rows(ase_depths, current_tibble)
```

Create index tibble
```{r}
cemt_table <- 
  tibble(cemt = list.files("/projects/epigenomics3/epigenomics3_results/users/ahauduc/arp/project/modules/1/output2/H3K4me3/", include.dirs = TRUE) %>% rep(6),
         individual_number = c(rep("1", 4), rep("2", 4), rep("3", 4), rep("4", 4), rep("5", 4), rep("6", 4), rep("7", 4), rep("8", 4)) %>% rep(6),
         mark = c(rep("H3K4me3", 8*4), rep("H3K27ac", 8*4), rep("H3K4me1", 8*4), rep("H3K9me3", 8*4), rep("H3K27me3", 8*4), rep("H3K36me3", 8*4)),
         celltype = rep(c("BC", "LP", "LC", "SC"), 192/4))
```

Loop for POSITIVE cis-hBTLs
Currently, this can't be run because some files are missing in folder ./ase/output
Must be run with a subsetted cemt_table tibble as in the jobs
```{r}
# Modules folder path
modules_path <- "/projects/epigenomics3/epigenomics3_results/users/ahauduc/arp/project/modules"
# Initialize results tibble
ase_depths <- tibble()

for (i in seq_along(cemt_table$cemt)) {
  # Import ChIP-seq treatment vcf file
  ase_treatment_vcf <- 
    readVcfAsVRanges(str_c(modules_path, "/notebook/alignment_analysis/ase/output/", cemt_table$mark[i], "_", cemt_table$celltype[i], "_", "Individual", "_", cemt_table$individual_number[i], "_treatment_cis_hbtls.vcf.gz")) %>%
    as.data.frame() %>%
    select(-width, -strand) %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE, ignore.strand = TRUE, seqinfo = seqinfo(BSgenome.Hsapiens.UCSC.hg38)) %>%
    keepSeqlevels(const_canonical_chromosomes, pruning.mode = "coarse")
  
  # Scan VCF for heterozygotic variant calls and keep only cis-hBTLs at heterozygotic positions for that individual
  # These are the positions that need to be tested for the particular individual-modality
  ase_het_positions <-
    join_overlap_intersect(consensus.all.spaas.rpkm.bed %>% 
                             plyranges::filter(mark == cemt_table$mark[i], celltype == cemt_table$celltype[i], spaa == TRUE) %>% 
                             keepSeqlevels(const_canonical_chromosomes, pruning.mode = "coarse"),
                           readVcfAsVRanges(str_c(modules_path, "/1/output2/", cemt_table$mark[i], "/", cemt_table$cemt[i], "/wgs/", cemt_table$cemt[i], ".recal_hard_filtered_phased_snvs.vcf.gz")) %>%
                             as.data.frame() %>%
                             select(-width, -strand) %>%
                             makeGRangesFromDataFrame(keep.extra.columns = TRUE, ignore.strand = TRUE, seqinfo = seqinfo(BSgenome.Hsapiens.UCSC.hg38)) %>%
                             keepSeqlevels(const_canonical_chromosomes, pruning.mode = "coarse") %>% 
                             plyranges::filter(AF == 0.5)) %>% 
    plyranges::select(-everything())
  
  # Subset the ChIP-seq treatment VCF with proper heterozygotic, modality-relevant positions
  # And select only the positions with some evidence of heterozygotic alignment
  ase_treatment_vcf_subsetted <-
    ase_treatment_vcf %>% 
    join_overlap_intersect(ase_het_positions) %>% 
    plyranges::filter(refDepth != 0 & altDepth != 0)
  
  # Pull out the vectors of depths where ref or alt != 0
  current_tibble <-
    tibble(mark = cemt_table$mark[i],
           celltype = cemt_table$celltype[i],
           individual = cemt_table$individual_number[i],
           ref_depth_mean = ase_treatment_vcf_subsetted$refDepth %>% mean(),
           alt_depth_mean = ase_treatment_vcf_subsetted$altDepth %>% mean())
  
  # Add to tibble
  ase_depths <- bind_rows(ase_depths, current_tibble)
}
```

# Combining intermediate results from the jobs separately run on each mark
```{r}
ase_depths <-
  bind_rows(ase_depths_H3K4me3,
            ase_depths_H3K27ac,
            ase_depths_H3K4me1,
            ase_depths_H3K9me3,
            ase_depths_H3K27me3,
            ase_depths_H3K36me3)
```

Visualization
```{r}
ase_depths %>% 
  mutate(`ALT / REF` = case_when(ref_depth_mean <  alt_depth_mean ~ "+",
                                 ref_depth_mean >= alt_depth_mean ~ "–") %>% factor(c("+", "–"))) %>%
  mutate(mark = mark %>% factor(vector_marks)) %>% 
  drop_na(`ALT / REF`) %>% 
  pivot_longer(c(ref_depth_mean, alt_depth_mean), names_to = "depth_mean_type", values_to = "depth") %>% 
  mutate(depth_mean_type = case_when(depth_mean_type == "ref_depth_mean" ~ "REF",
                                     depth_mean_type == "alt_depth_mean" ~ "ALT")) %>% 
  mutate(depth_mean_type = depth_mean_type %>% factor(c("REF", "ALT"))) %>% 
  mutate(id = str_c(mark, celltype, individual)) %>% 
  ggplot(aes(x = depth_mean_type, y = depth)) +
  geom_line(aes(group = id, color = `ALT / REF`), size = 0.2) +
  scale_colour_manual(values = c(`+` = "#ff9999", `–` = "#b3d7ff")) +
  geom_point(size = 0.7) +
  xlab(element_blank()) +
  ylab("Mean ChIP Treatment Read Allele Depth\nat Heterozygous Alt-Allele-Driven cis-hBTLs") + 
  theme(axis.title.y = element_text(angle = 0)) +
  theme_bw() + 
  facet_grid(cols = vars(mark))
ggsave("plots/dotplot_ase_mean_allele_counts.png",
       units = "in",
       width = 10,
       height = 5)
```
