---
title: "ROI Analysis"
output: html_notebook
---
```{r}
# R 4.1.3
# x86_64-centos7-linux-gnu
library(plyranges)
library(VariantAnnotation)
library(MutationalPatterns)
library(genomation)
library(Gviz)
library(BSgenome.Hsapiens.UCSC.hg38)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(systemPipeR)
library(biomaRt)
library(LDlinkR)
library(AllelicImbalance)
library(broom)
library(gt)
library(gtsummary)
library(UpSetR)
library(tidyverse)

const_canonical_chromosomes <- str_c("chr", c(1:22, "X", "Y"))

vector_marks <- c("H3K4me3", "H3K4me1", "H3K27ac", "H3K9me3", "H3K27me3", "H3K36me3")
vector_celltypes <- c("BC", "LP", "LC", "SC")
vector_individuals <- c("Individual_14_17_LP_DNA",
                        "Individual_14_18_BC__356K_RNA",
                        "Individual_11_18_BC_RNA_DNA",
                        "Individual_22_18_BC_RNA_DNA",
                        "Individual_24_18_BC_RNA_DNA",
                        "Individual_38_18_LC_RNA_DNA",
                        "Individual_30_18_LC_RNA_DNA",
                        "Individual_15_18_LC_RNA_DNA")


```

Read in VCF
```{r}
top_hits_mcf10a_parental <- 
  readVcfAsVRanges("roi_alignments/output/mcf10a_parental/all_rois.vcf.gz") %>%
  as.data.frame() %>%
  select(-width, -strand) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE, ignore.strand = TRUE, seqinfo = seqinfo(BSgenome.Hsapiens.UCSC.hg38))
```

Compare VCF with know top hits
Only the variant that was ALT in MCF10A actually shows up as alt in the variant calls
None of the other variants show up because those are supposed to be wildtype
```{r}
join_overlap_intersect(top_hits_ranges, top_hits_mcf10a_parental)
```

Read in VCF
```{r}
top_hits_mcf10a_edit_rn1 <- 
  readVcfAsVRanges("roi_alignments/output/mcf10a_edit_rn1/all_rois.vcf.gz") %>%
  as.data.frame() %>%
  select(-width, -strand) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE, ignore.strand = TRUE, seqinfo = seqinfo(BSgenome.Hsapiens.UCSC.hg38))
```

Compare VCF with know top hits
Only the variant that was ALT in MCF10A actually shows up as alt in the variant calls
None of the other variants show up because those are supposed to be wildtype
```{r}
join_overlap_intersect(top_hits_ranges, top_hits_mcf10a_edit_rn1)
```


Visualize the genomic alignments
```{r}
# Set reusable variables
gviz_genome <- "hg38"
gviz_cyto_bands <- read.table("data/hg38_cytoBandIdeo.txt", header = FALSE, sep = "\t", col.names = c("chrom", "chromStart", "chromEnd", "name", "gieStain"))
gviz_biomart_hg38_genes <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
```

Plot any locus
```{r}
# Plot any locus
plot_top_hits_any_locus <- function(bampath, chr, start, bp_left, bp_right) {
  plotTracks(chromosome = chr,
             from =       start + bp_left, 
             to =         start + bp_right,
             transcriptAnnotation = "symbol", 
             stacking = "squish",
             #type = "coverage",
             #####################################################################
             # Data Track List
             trackList = list(
               # Ideogram
               IdeogramTrack(chromosome = chr,
                             genome = gviz_genome,
                             bands = gviz_cyto_bands),
               # Genome Axis
               GenomeAxisTrack(),
               
               ##### Cis-hBTL markings
               AnnotationTrack(top_hits_ranges,
                               name = "cis-hBTLs",
                               genome = gviz_genome,
                               col = "red"),
               # Read alignments
               AlignmentsTrack(bampath,
                               isPaired = TRUE),
               SequenceTrack(Hsapiens, 
                             chromosome = chr),
               
               ### Gene regions
               BiomartGeneRegionTrack(biomart = gviz_biomart_hg38_genes,
                                      genome = gviz_genome,
                                      # symbol = "MMP7",
                                      name = "ENSEMBL Genes",
                                      filter = list(with_refseq_mrna = TRUE))))
}


```


Plot cohort + cell lines
```{r}
# Plot cohort + cell lines
plot_top_hits <- function(bampath, rsID, coordLeft, coordRight) {
  plotTracks(chromosome = top_hits_ranges %>% plyranges::filter(rs == rsID) %>% slice(1) %>% seqnames() %>% as.character(),
             from =       top_hits_ranges %>% plyranges::filter(rs == rsID) %>% slice(1) %>% start() %>% `+`(coordLeft), 
             to =         top_hits_ranges %>% plyranges::filter(rs == rsID) %>% slice(1) %>% start() %>% `+`(coordRight),
             transcriptAnnotation = "symbol", 
             stacking = "squish",
             #type = "coverage",
             #####################################################################
             # Data Track List
             trackList = list(
               # Ideogram
               IdeogramTrack(chromosome = top_hits_ranges %>% plyranges::filter(rs == rsID) %>% slice(1) %>% seqnames() %>% as.character(),
                             genome = gviz_genome,
                             bands = gviz_cyto_bands),
               # Genome Axis
               GenomeAxisTrack(),
               
               ##### Cis-hBTL markings
               AnnotationTrack(top_hits_ranges,
                               name = "cis-hBTLs",
                               genome = gviz_genome,
                               col = "red"),
               # Read alignments
               AlignmentsTrack(bampath,
                               isPaired = TRUE),
               SequenceTrack(Hsapiens, 
                             chromosome = top_hits_ranges %>% plyranges::filter(rs == rsID) %>% slice(1) %>% seqnames() %>% as.character()),
               
               ### Gene regions
               BiomartGeneRegionTrack(biomart = gviz_biomart_hg38_genes,
                                      genome = gviz_genome,
                                      # symbol = "MMP7",
                                      name = "ENSEMBL Genes",
                                      filter = list(with_refseq_mrna = TRUE))))
}
```

Plot mcf10a_parental
```{r}
# 0 ################################################################################
# Control locus ABCA1 +/- 20
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs75071948", "_", "pm20", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits_any_locus(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", chr = "chr9", start = 104960001, bp_left = -20, bp_right = 20)
dev.off()

# Control locus ABCA1 +/- 200
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs75071948", "_", "pm200", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits_any_locus(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", chr = "chr9", start = 104960001, bp_left = -200, bp_right = 200)
dev.off()

# 1 ################################################################################
# rs75071948 +/- 20
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs75071948", "_", "pm20", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs75071948", -20, 20)
dev.off()

# rs75071948 +/- 200
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs75071948", "_", "pm200", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs75071948", -200, 200)
dev.off()

# 2 ###############################################################################
# rs10502002 +/- 20
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs10502002", "_", "pm20", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs10502002", -20, 20)
dev.off()

# rs10502002 +/- 200
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs10502002", "_", "pm200", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs10502002", -200, 200)
dev.off()

# 3 ###############################################################################
# rs1879068 +/- 20
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs1879068", "_", "pm20", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs1879068", -20, 20)
dev.off()

# rs1879068 +/- 200
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs1879068", "_", "pm200", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs1879068", -200, 200)
dev.off()

# 4 ###############################################################################
# rs4647960 +/- 20
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs4647960", "_", "pm20", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs4647960", -20, 20)
dev.off()

# rs4647960 +/- 200
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs4647960", "_", "pm200", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs4647960", -200, 200)
dev.off()

# 5 ###############################################################################
# rs72835059 +/- 20
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs72835059", "_", "pm20", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs72835059", -20, 20)
dev.off()

# rs72835059 +/- 200
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs72835059", "_", "pm200", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs72835059", -200, 200)
dev.off()

# 6 ##############################################################################
# rs553562345 +/- 20
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs553562345", "_", "pm20", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs553562345", -20, 20)
dev.off()

# rs553562345 +/- 200
png(str_c("plots_gviz/", "top_hits_mcf10a_parental_", "rs553562345", "_", "pm200", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_parental/all_rois.dedup.bam", "rs553562345", -200, 200)
dev.off()
```

Plot mcf10a_edit_rn1
```{r}
# 0 ################################################################################
# Control locus ABCA1 +/- 20
png(str_c("plots_gviz/", "top_hits_mcf10a_edit_rn1_", "rs4743771", "_", "pm20", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits_any_locus(bampath = "roi_alignments/output/mcf10a_edit_rn1/all_rois.dedup.bam", chr = "chr9", start = 104960001, bp_left = -20, bp_right = 20)
dev.off()

# Control locus ABCA1 +/- 200
png(str_c("plots_gviz/", "top_hits_mcf10a_edit_rn1_", "rs75071948", "_", "pm200", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits_any_locus(bampath = "roi_alignments/output/mcf10a_edit_rn1/all_rois.dedup.bam", chr = "chr9", start = 104960001, bp_left = -200, bp_right = 200)
dev.off()

# 1 ################################################################################
# rs75071948 +/- 20
png(str_c("plots_gviz/", "top_hits_mcf10a_edit_rn1_", "rs75071948", "_", "pm20", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_edit_rn1/all_rois.dedup.bam", "rs75071948", -20, 20)
dev.off()

# rs75071948 +/- 200
png(str_c("plots_gviz/", "top_hits_mcf10a_edit_rn1_", "rs75071948", "_", "pm200", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_edit_rn1/all_rois.dedup.bam", "rs75071948", -200, 200)
dev.off()

# 2 ###############################################################################
# rs10502002 +/- 20
png(str_c("plots_gviz/", "top_hits_mcf10a_edit_rn1_", "rs10502002", "_", "pm20", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_edit_rn1/all_rois.dedup.bam", "rs10502002", -20, 20)
dev.off()

# rs10502002 +/- 200
png(str_c("plots_gviz/", "top_hits_mcf10a_edit_rn1_", "rs10502002", "_", "pm200", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_edit_rn1/all_rois.dedup.bam", "rs10502002", -200, 200)
dev.off()

# 3 ###############################################################################
# rs1879068 +/- 20
png(str_c("plots_gviz/", "top_hits_mcf10a_edit_rn1_", "rs1879068", "_", "pm20", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_edit_rn1/all_rois.dedup.bam", "rs1879068", -20, 20)
dev.off()

# rs1879068 +/- 200
png(str_c("plots_gviz/", "top_hits_mcf10a_edit_rn1_", "rs1879068", "_", "pm200", ".png"),
    width = 14,
    height = 10,
    units = "in",
    res = 300)
plot_top_hits(bampath = "roi_alignments/output/mcf10a_edit_rn1/all_rois.dedup.bam", "rs1879068", -200, 200)
dev.off()
```

