---
title: "Total variants and cis-hBTLs Venn diagrams"
output: html_notebook
---
```{r}
#!/gsc/software/linux-x86_64-centos7/R-4.1.3/lib64/R/bin/R
# R 4.1.3
# x86_64-centos7-linux-gnu
library(plyranges)
library(VariantAnnotation)
library(MutationalPatterns)
library(genomation)
library(Gviz)
library(BSgenome.Hsapiens.UCSC.hg38)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(systemPipeR)
library(biomaRt)
library(LDlinkR)
library(AllelicImbalance)
library(broom)
library(gt)
library(gtsummary)
library(UpSetR)
library(nVennR)
library(tidyverse)
library(ggdendro)

const_canonical_chromosomes <- str_c("chr", c(1:22, "X"))

vector_marks <- c("H3K4me3", "H3K4me1", "H3K27ac", "H3K9me3", "H3K27me3", "H3K36me3")
vector_celltypes <- c("BC", "LP", "LC", "SC")
vector_individuals <- c("Individual_14_17_LP_DNA",
                        "Individual_14_18_BC__356K_RNA",
                        "Individual_11_18_BC_RNA_DNA",
                        "Individual_22_18_BC_RNA_DNA",
                        "Individual_24_18_BC_RNA_DNA",
                        "Individual_38_18_LC_RNA_DNA",
                        "Individual_30_18_LC_RNA_DNA",
                        "Individual_15_18_LC_RNA_DNA")

vector_individuals_encoded <- c("Individual_1",
                                "Individual_2",
                                "Individual_3",
                                "Individual_4",
                                "Individual_5",
                                "Individual_6",
                                "Individual_7",
                                "Individual_8")
```

1. Read in total variants across cohort and merge them together into a single GRanges (variant_calls_merged)
```{r}
variant_calls_merged <- 
  list_variant_calls %>% 
  unlist() %>% 
  reduce_ranges() %>% 
  keepSeqlevels(const_canonical_chromosomes, pruning.mode = "coarse")

```

2. Annotate merged variant calls with cis-hBTL status by celltype
```{r}
variant_calls_merged_annotated <-
  variant_calls_merged %>% 
  plyranges::mutate(key = str_c(seqnames(.), "_", start(.), "_", end(.)),
                    all = TRUE,
                    cis_hBTL_any = countOverlaps(., consensus.all.spaas.rpkm.bed) > 0,
                    cis_hBTL_BC = countOverlaps(., consensus.all.spaas.rpkm.bed %>% plyranges::filter(celltype == "BC")) > 0,
                    cis_hBTL_LP = countOverlaps(., consensus.all.spaas.rpkm.bed %>% plyranges::filter(celltype == "LP")) > 0,
                    cis_hBTL_LC = countOverlaps(., consensus.all.spaas.rpkm.bed %>% plyranges::filter(celltype == "LC")) > 0,
                    cis_hBTL_SC = countOverlaps(., consensus.all.spaas.rpkm.bed %>% plyranges::filter(celltype == "SC")) > 0)

```

3. Create Venn-ready list
```{r}
venn_cis_hBTL_list <- 
  list(`All Variants` = variant_calls_merged_annotated %>% `$`(key), 
       `BC cis-hBTLs`  = variant_calls_merged_annotated %>% plyranges::filter(cis_hBTL_BC == TRUE) %>% `$`(key), 
       `LP cis-hBTLs`  = variant_calls_merged_annotated %>% plyranges::filter(cis_hBTL_LP == TRUE) %>% `$`(key),
       `LC cis-hBTLs`  = variant_calls_merged_annotated %>% plyranges::filter(cis_hBTL_LC == TRUE) %>% `$`(key),
       `SC cis-hBTLs`  = variant_calls_merged_annotated %>% plyranges::filter(cis_hBTL_SC == TRUE) %>% `$`(key))
```

4. Graph Venn object with nVennR
```{r}
venn_cis_hBTL_list %>% 
  plotVenn() %>% 
  showSVG(outFile = str_c("plots/", "venn_nvennr_all_variants_celltype_cis_hBTLs.svg"), 
          opacity = 0.1, 
          fontScale = 0.8,
          labelRegions = FALSE,
          setColors = c("#d3d3d3",  # All- Light Gray
                        "#3a68ae",  # BC - Blue
                        "#ff3739",  # LP - Red
                        "#eb8353",  # LC - Orange
                        "#464546")) # SC - Dark Gray
```
