---
title: "ANXA1 RPKM Analysis"
output: html_notebook
---
```{r}
#!/gsc/software/linux-x86_64-centos7/R-4.1.3/lib64/R/bin/R
# R 4.1.3
# x86_64-centos7-linux-gnu
library(plyranges)
library(VariantAnnotation)
library(MutationalPatterns)
library(genomation)
library(Gviz)
library(BSgenome.Hsapiens.UCSC.hg38)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(systemPipeR)
library(biomaRt)
library(LDlinkR)
library(AllelicImbalance)
library(broom)
library(gt)
library(gtsummary)
library(UpSetR)
library(tidyverse)

const_canonical_chromosomes <- str_c("chr", c(1:22, "X"))

vector_marks <- c("H3K4me3", "H3K27ac", "H3K4me1", "H3K9me3", "H3K27me3", "H3K36me3")
vector_celltypes <- c("BC", "LP", "LC", "SC")
vector_individuals <- c("Individual_14_17_LP_DNA",
                        "Individual_14_18_BC__356K_RNA",
                        "Individual_11_18_BC_RNA_DNA",
                        "Individual_22_18_BC_RNA_DNA",
                        "Individual_24_18_BC_RNA_DNA",
                        "Individual_38_18_LC_RNA_DNA",
                        "Individual_30_18_LC_RNA_DNA",
                        "Individual_15_18_LC_RNA_DNA")


```

Mean number of variants per individual
```{r}
list_variant_calls_length <- list()

for (i in 1:8) {
  list_variant_calls_length[[i]] <- list_variant_calls[[i]] %>% as_tibble() %>% nrow()
}

list_variant_calls_length %>% as_vector() %>% mean()
```

Count total dbSNP all and common variants
```{r}
# Load in all snps and only keep the ones that are within all the variants called on breast in your pipeline
vcf_snps_all <- 
  readVcfAsVRanges("/projects/epigenomics3/epigenomics3_results/users/ahauduc/arp/genome_data/dbSNP/snps/00-All_chr.snps.vcf.gz") %>%
  as.data.frame() %>%
  select(-width, -strand) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE, ignore.strand = TRUE) %>% 
  join_overlap_intersect(list_variant_calls %>% unlist() %>% reduce_ranges())

# Load in all snps and only keep the ones that are within all the variants called on breast in your pipeline
vcf_snps_common <- 
  readVcfAsVRanges("/projects/epigenomics3/epigenomics3_results/users/ahauduc/arp/genome_data/dbSNP/common_snps/00-common_snps_chr.vcf.gz") %>%
  as.data.frame() %>%
  select(-width, -strand) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE, ignore.strand = TRUE) %>% 
  join_overlap_intersect(list_variant_calls %>% unlist() %>% reduce_ranges())
```

```{r}
# First, view non-redundant variant call ranges from breast cohort
# Then, left join them with dbSNP GRanges
# Then, count the number of your variants that have a corresponding dbSNP variant by leveraging the RV column (dbSNP table reduced to only keep unique positions)
# All dbSNP SNVs
cis_hBTL_present_in_dbsnp_all <- 
  list_variant_calls %>% 
  unlist() %>% 
  reduce_ranges() %>% 
  join_overlap_left(vcf_snps_all %>% reduce_ranges() %>% plyranges::mutate(dbSNP_present = TRUE)) %>% 
  plyranges::mutate(cis_hBTL_present_in_dbsnp_set = case_when(!is.na(dbSNP_present) ~ TRUE,
                                                              is.na(dbSNP_present) ~ FALSE)) %>% 
  keepSeqlevels(const_canonical_chromosomes, pruning.mode = "coarse") %>% 
  `$`(cis_hBTL_present_in_dbsnp_set) %>% 
  table() %>% 
  as_tibble() %>% 
  mutate(dbSNP_snv_set = "all") %>% 
  `colnames<-`(c("present_in_cis_hBTL_set", "number", "dbSNP_snv_set"))

# Repeat for common dbSNP SNVs
cis_hBTL_present_in_dbsnp_common <- 
  list_variant_calls %>% 
  unlist() %>% 
  reduce_ranges() %>% 
  join_overlap_left(vcf_snps_common %>% reduce_ranges() %>% plyranges::mutate(dbSNP_present = TRUE)) %>% 
  plyranges::mutate(cis_hBTL_present_in_dbsnp_set = case_when(!is.na(dbSNP_present) ~ TRUE,
                                                              is.na(dbSNP_present) ~ FALSE)) %>% 
  keepSeqlevels(const_canonical_chromosomes, pruning.mode = "coarse") %>% 
  `$`(cis_hBTL_present_in_dbsnp_set) %>% 
  table() %>% 
  as_tibble() %>% 
  mutate(dbSNP_snv_set = "common") %>% 
  `colnames<-`(c("present_in_cis_hBTL_set", "number", "dbSNP_snv_set"))
```

```{r}
# Create output tibble from above data, and get percentages for every dbSNP set
dbSNP_set_cis_hBTL_intersects <- 
  bind_rows(cis_hBTL_present_in_dbsnp_all,
            cis_hBTL_present_in_dbsnp_common) %>% 
  group_by(dbSNP_snv_set) %>% 
  mutate(fraction_in_set = number/sum(number)) %>% 
  ungroup()

# Plot
dbSNP_set_cis_hBTL_intersects %>% 
  ggplot(aes(x = dbSNP_snv_set, y = number, fill = present_in_cis_hBTL_set %>% factor(c(TRUE, FALSE)), label = scales::percent(fraction_in_set, accuracy = 0.01))) +
  geom_bar(position = "dodge", stat = "identity") +
  scale_fill_manual(values = c("black", "light gray")) +
  geom_text(position = position_dodge(width = .9),    # move to center of bars
              vjust = -0.5,    # nudge above top of bar
              size = 3) +
  labs(x = "dbSNP set", y = "Cis-hBTL loci", fill = "Cis-hBTL locus in dbSNP set") +
  theme_bw()
ggsave("plots/bargraph_dbSNP_all_vs_common.png",
       units = "in",
       width = 7,
       height = 4)

```

Calculate zygosity
```{r}

```




