---
title: "Redo of SPAA Analysis including negative SPAAs"
output: html_notebook
---
```{r}
#!/gsc/software/linux-x86_64-centos7/R-4.1.3/lib64/R/bin/R
# R 4.1.3
# x86_64-centos7-linux-gnu
library(plyranges)
library(VariantAnnotation)
library(MutationalPatterns)
library(genomation)
library(Gviz)
library(BSgenome.Hsapiens.UCSC.hg38)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(systemPipeR)
library(biomaRt)
library(LDlinkR)
library(AllelicImbalance)
library(broom)
library(gt)
library(gtsummary)
library(UpSetR)
library(tidyverse)
library(ggdendro)

const_canonical_chromosomes <- str_c("chr", c(1:22, "X"))

vector_marks <- c("H3K4me3", "H3K27ac", "H3K4me1", "H3K9me3", "H3K27me3", "H3K36me3")
vector_celltypes <- c("BC", "LP", "LC", "SC")
vector_individuals <- c("Individual_14_17_LP_DNA",
                        "Individual_14_18_BC__356K_RNA",
                        "Individual_11_18_BC_RNA_DNA",
                        "Individual_22_18_BC_RNA_DNA",
                        "Individual_24_18_BC_RNA_DNA",
                        "Individual_38_18_LC_RNA_DNA",
                        "Individual_30_18_LC_RNA_DNA",
                        "Individual_15_18_LC_RNA_DNA")

vector_individuals_encoded <- c("Individual_1",
                                "Individual_2",
                                "Individual_3",
                                "Individual_4",
                                "Individual_5",
                                "Individual_6",
                                "Individual_7",
                                "Individual_8")
```

# Plot SPAAs
```{r}
consensus.all.spaas.rpkm.bed %>% 
  as_tibble() %>% 
  group_by(individuals_involved, spaa) %>% 
  summarize(number_of_spaas = n()) %>% 
  mutate(`Cis-hBTL Type` = if_else(spaa == TRUE, "Alt Allele-Driven", "Ref Allele-Driven")) %>% 
  ggplot(aes(x = as.factor(individuals_involved), y = number_of_spaas, fill = `Cis-hBTL Type`)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_bw() +
  scale_fill_grey() +
  labs(x = "Number of Individuals Involved", y = "Number of cis-hBTLs")
ggsave("plots/barplot_individuals_involved.png",
       units = "in",
       width = 6,
       height = 3.5)
```

Old
```{r}
# consensus.all.spaas.rpkm.bed %>% 
#   as_tibble() %>%
#   group_by(spaa) %>% 
#   summarize(number_of_spaas = n()) %>% 
#   mutate(`Cis-hBTL Type` = if_else(spaa == TRUE, "alt_allele_driven", "ref_allele_driven")) %>% 
#   ggplot(aes(x = "", y = number_of_spaas, fill = `Cis-hBTL Type`)) +
#   geom_bar(width = 1, stat = "identity") + 
#   coord_polar("y", start = 0) +
#   labs(x = "", y = "")
# 
# consensus.all.spaas.rpkm.bed %>% 
#   as_tibble() %>%
#   filter(t_test_p_val_fdr_adj < 0.01) %>% 
#   group_by(spaa) %>% 
#   summarize(number_of_spaas = n()) %>% 
#   mutate(`Cis-hBTL Type` = if_else(spaa == TRUE, "alt_allele_driven", "ref_allele_driven")) %>% 
#   ggplot(aes(x = "", y = number_of_spaas, fill = `Cis-hBTL Type`)) +
#   geom_bar(width = 1, stat = "identity") + 
#   coord_polar("y", start = 0) +
#   labs(x = "", y = "")

# consensus.all.spaas.rpkm.bed %>% 
#   as_tibble() %>% 
#   filter(t_test_p_val_fdr_adj < 0.01) %>% 
#   group_by(individuals_involved, spaa) %>% 
#   summarize(number_of_spaas = n()) %>% 
#   mutate(`Cis-hBTL Type` = if_else(spaa == TRUE, "alt_allele_driven", "ref_allele_driven")) %>% 
#   ggplot(aes(x = as.factor(individuals_involved), y = number_of_spaas, fill = `Cis-hBTL Type`)) +
#   geom_bar(stat = "identity", position = position_dodge()) +
#   labs(x = "Number of Individuals Involved", y = "Number of cis-hBTLs")
```
