---
title: "Amplicon Analysis"
output: html_notebook
---
```{r}
# R 4.1.3
# x86_64-centos7-linux-gnu
library(plyranges)
library(VariantAnnotation)
library(MutationalPatterns)
library(genomation)
library(Gviz)
library(BSgenome.Hsapiens.UCSC.hg38)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(systemPipeR)
library(biomaRt)
library(LDlinkR)
library(AllelicImbalance)
library(broom)
library(gt)
library(gtsummary)
library(UpSetR)
library(tidyverse)

const_canonical_chromosomes <- str_c("chr", c(1:22, "X", "Y"))

vector_marks <- c("H3K4me3", "H3K4me1", "H3K27ac", "H3K9me3", "H3K27me3", "H3K36me3")
vector_celltypes <- c("BC", "LP", "LC", "SC")
vector_individuals <- c("Individual_14_17_LP_DNA",
                        "Individual_14_18_BC__356K_RNA",
                        "Individual_11_18_BC_RNA_DNA",
                        "Individual_22_18_BC_RNA_DNA",
                        "Individual_24_18_BC_RNA_DNA",
                        "Individual_38_18_LC_RNA_DNA",
                        "Individual_30_18_LC_RNA_DNA",
                        "Individual_15_18_LC_RNA_DNA")

```

Count the number of lines containing a perfect edited haplotype in the fastq.gz files using system
```{r}
# Set constants
validation_file_path_base <- "/projects/epigenomics3/epigenomics3_results/users/ahauduc/arp/project/modules/notebook/alignment_analysis/roi_alignments/data/batch_2/dirs/"
validation_perfect_hdr_haplotype <- "CCAATGGAAAGTGATAATACAGAAAGGTGGGGCTGGTGTAGGGGTGAAGGCAGGATGCTTTGGGAGAGCATGGAAGGTCACC"

# Create vector of batch 2 IDs and table to receive data
validation_batch_2_ids <- read_csv("data/validation/batch_2_ids.csv", col_names = FALSE) %>% pull()
validation_batch_2_info_raw <- tibble()

# Loop through fastq files and tally number of reads containing perfect hdr haplotype
for (i in seq_along(validation_batch_2_ids)) {
  # Run bash command to tally number of reads with correct HDR-edited haplotype
  tmp_df <- 
    tibble(
      id = 
        validation_batch_2_ids[i],
      number_reads_w_perfect_hdr = 
        system(str_c("zcat ", validation_file_path_base, validation_batch_2_ids[i], "/100bp/merge_fastq/*1_100bp_4_lanes.merge.fastq.gz | grep ", validation_perfect_hdr_haplotype, " | wc -l"), 
               intern = TRUE) %>% 
        as.integer(),
      number_reads_total = 
        system(str_c("zcat ", validation_file_path_base, validation_batch_2_ids[i], "/100bp/merge_fastq/*1_100bp_4_lanes.merge.fastq.gz | wc -l"), 
               intern = TRUE) %>% 
        as.integer() %>% 
        `/`(4) # divide number of fastq lines by 4 to get total number of reads
        )
  
  # Add to new row of receiving dataframe
  validation_batch_2_info_raw <- bind_rows(validation_batch_2_info_raw, tmp_df)
  
  # Remove temporary dataframe
  rm(tmp_df)
  
  # Keep track
  message("Done with ", validation_batch_2_ids[i])
}
```

Read genomic alignments
```{r}
# Top hit ANXA1 range
validation_region_anxa1 <- top_hits_ranges %>% plyranges::filter(rs == "rs75071948") %>% reduce_ranges() %>% keepSeqlevels("chr9", pruning.mode = "coarse")

# Loop through bases and add allele counts at rs75071948 for all samples
for (BASE in c("A", "C", "G", "T")) {
  tmp_allele_counts <- 
    countAllelesFromBam(
      gr = validation_region_anxa1,
      pathToDir = "/projects/epigenomics3/epigenomics3_results/users/ahauduc/arp/project/modules/notebook/alignment_analysis/roi_alignments/output/batch_2/alignments/")[,,BASE] %>% 
    as.data.frame() %>% 
    rownames_to_column("id") %>% 
    mutate(id = id %>% str_remove(".dedup.bam"), .before = 1) %>% 
    `colnames<-`(c("id", str_c("number_unique_reads_allele_", BASE)))
  
  # Add to info for cohort
  validation_batch_2_info_raw <- left_join(validation_batch_2_info_raw, tmp_allele_counts, by = "id")
  
  # Remove tmp file
  rm(tmp_allele_counts)
}
```

Process validation_batch_2_info_raw
```{r}
# Calculate read fractions
validation_batch_2_info_processed <- 
  validation_batch_2_info_raw %>% 
  mutate(`Perfect HDR`    = number_reads_w_perfect_hdr   / number_reads_total,
         `Correct allele` = (number_unique_reads_allele_G / (number_unique_reads_allele_A + number_unique_reads_allele_C + number_unique_reads_allele_G + number_unique_reads_allele_T) - `Perfect HDR`), #subtracting perfect HDR to not double-count reads with perfect haplotype
         Other            = 1 - (`Perfect HDR` + `Correct allele`)) %>% 
  pivot_longer(cols = c(`Perfect HDR`, `Correct allele`, Other),
               names_to = "Read type",
               values_to = "fraction")

# Plot as multi-Venn
validation_batch_2_info_processed %>% 
  mutate(id = id %>% factor(validation_batch_2_ids),
         `Read type` = `Read type` %>% factor(c("Perfect HDR", "Correct allele", "Other"))) %>% 
  ggplot(aes(x = "", y = fraction, fill = `Read type`)) +
  geom_bar(stat = "identity", position = position_fill()) +
  coord_polar(theta = "y") +
  facet_wrap(~id) +
  theme(axis.title.x = element_blank(),
        axis.title.y = element_blank()) + 
  scale_fill_grey() +
  theme_void()
ggsave("plots/piechart_validation_read_types_batch2.png")
```
